### ISSUE TITLE
Surface Pro 4: Battery sometimes still set as AC power even when discharging

### ISSUE THREAD

**Issue by battaglia01** [jakeday/linux-surface#223]
I'm running KDE Neon (based on Ubuntu Xenial). Sometimes when I unplug my battery, the system will still think the AC adapter is plugged in, although the battery status does get updated as "discharging." I end up in a strange state where the AC adapter is in the "online" state, yet the battery is in the "discharging" state.

The problem does not seem to be limited to KDE. Rather, the issue seems to be that `upower` and `acpi` don't agree. `acpi` seems to give the right results and `upower` seems to be wrong.

For example, below, the system is a state where the icon still (erroneously) says it's plugged in, but the battery is discharging in real life:

```
~ $ acpi -V
Battery 0: Discharging, 52%, 02:05:19 remaining
Battery 0: design capacity 4936 mAh, last full capacity 4036 mAh = 81%
Adapter 0: off-line
Thermal 0: ok, 41.0 degrees C
Thermal 0: trip point 0 switches to mode hot at temperature 240.0 degrees C
Thermal 1: ok, 36.4 degrees C
Thermal 1: trip point 0 switches to mode hot at temperature 60.0 degrees C
Thermal 2: ok, 0.0 degrees C
Thermal 2: trip point 0 switches to mode hot at temperature 97.0 degrees C
Thermal 3: ok, 40.0 degrees C
Thermal 3: trip point 0 switches to mode hot at temperature 54.0 degrees C
Cooling 0: x86_pkg_temp no state information available
Cooling 1: Processor 0 of 10
Cooling 2: Processor 0 of 10
Cooling 3: intel_powerclamp no state information available
Cooling 4: pch_skylake no state information available
Cooling 5: Processor 0 of 10
Cooling 6: Processor 0 of 10
```
So far so good - note that Battery 0 is discharging and Adapter 0 is offline - as it should be. On the other hand, here's upower:

```
$ upower -d
Device: /org/freedesktop/UPower/devices/line_power_ADP1
  native-path:          ADP1
  power supply:         yes
  updated:              Thu 05 Jul 2018 12:54:02 AM EDT (2586 seconds ago)
  has history:          no
  has statistics:       no
  line-power
    warning-level:       none
    online:              yes
    icon-name:          'ac-adapter-symbolic'

Device: /org/freedesktop/UPower/devices/battery_BAT1
  native-path:          BAT1
  vendor:               SMP
  model:                X910527
  serial:               39135
  power supply:         yes
  updated:              Thu 05 Jul 2018 01:36:28 AM EDT (40 seconds ago)
  has history:          yes
  has statistics:       yes
  battery
    present:             yes
    rechargeable:        yes
    state:               discharging
    warning-level:       none
    energy:              16.387 Wh
    energy-empty:        0 Wh
    energy-full:         31.192 Wh
    energy-full-design:  38.152 Wh
    energy-rate:         7.747 W
    voltage:             7.749 V
    time to empty:       2.1 hours
    percentage:          52%
    capacity:            81.7572%
    technology:          lithium-ion
    icon-name:          'battery-good-symbolic'
  History (charge):
    1530768986	52.000	charging
  History (rate):
    1530768988	7.747	discharging
    1530768986	18.727	charging

Device: /org/freedesktop/UPower/devices/DisplayDevice
  power supply:         yes
  updated:              Thu 05 Jul 2018 01:36:28 AM EDT (40 seconds ago)
  has history:          no
  has statistics:       no
  battery
    present:             yes
    state:               discharging
    warning-level:       none
    energy:              16.387 Wh
    energy-full:         31.192 Wh
    energy-rate:         7.747 W
    time to empty:       2.1 hours
    percentage:          52%
    icon-name:          'battery-good-symbolic'

Daemon:
  daemon-version:  0.99.4
  on-battery:      no
  lid-is-closed:   no
  lid-is-present:  yes
  critical-action: HybridSleep
```

Note at the very top, it thinks the AC adapter (ADP1) is still online! Note the online: yes

Any clue what might cause this?

**Comment by m9115m** [comment]
I've seen the opposite on a SB4...
STR
1. Charge battery (less than 100%)
2. Unplug power (you see the battery icon, and note that you are discharging)
3. Plug power back in (whaa.... you still see the battery icon, and still says you are discharging)

Perhaps some other SB4 fans could confirm?

As stated in the OP, there seems to be a "disconnect" (sorry) between various power status states.

One other interesting note, after a few minutes, my SB4 (still under 100%) figured out that it was charging again - perhaps there's a refresh issue thrown in for fun...

Shameless plug.  This build rocks!

**Comment by battaglia01** [comment]
I was able to fix this. The problem is that sometimes the "battery on/off" message isn't sent correctly to upower. ACPI still works, so this is an issue with upower.

To fix this, I added a custom ACPI script in /etc/acpi/battery.sh

```
#!/bin/sh
sleep 1
busctl call --system org.freedesktop.UPower /org/freedesktop/UPower/devices/line_power_ADP1 org.freedesktop.UPower.Device Refresh
busctl call --system org.freedesktop.UPower /org/freedesktop/UPower/devices/battery_BAT1 org.freedesktop.UPower.Device Refresh
```

which just sleeps for one second and then refreshes the battery and AC status. Then I added an ACPI event to call the above script any time a upon power connect/disconnect  happens to /etc/acpi/events/battery

```
# /etc/acpi/events/battery
# This is called when the user presses the power button and calls
# /etc/acpi/battery.sh for further processing.

# Optionally you can specify the placeholder %e. It will pass
# through the whole kernel event message to the program you've
# specified.

event=battery.*
action=/etc/acpi/battery.sh
```

If you do the above, you will likely need to restart to see changes take place.

**Comment by m9115m** [comment]
Tried this on Xubuntu 18.04.1. Kernel 17.3. Didn't work. :(

Seems like @battaglia01 is on the right track.  There seems to be an issue between the polling of battery status (am I plugged, on battery, and percentage), and acpi plugged/unplugged events.

Random guess, there are overlapping events and perhaps... an assumption in code that if I'm plugged and get an unplugged event, I assume unplugged status (and vica versa).

The problem when you assume... (for all you Bad News Bears fans out there)

If the plug/unplug occurs too quickly, and assumptions are made that supercede polling, then it seems possible to get out of sync.  (i.e. I'm plugged, but showing an unplugged icon)

Interesting that @battaglia01 got it working in KDE.  Likely the power polling/events are different in Gnome.

Bottom line, seems like there's a lower level issue with the way power status and events are reported to the kernel on SP4.

Wondering @jakeday is the acpi hardware be different on a SP4, such that the events/polling is getting out of sync?

**Comment by battaglia01** [comment]
I'm on the v16 kernel, so you may want to try downgrading to see if that
does it for you. But I'm not quite sure what protocol GNOME uses,
unfortunately. I also used some custom config settings when compiling, so
I'm not sure if that's involved.

**Comment by jakeday** [comment]
This has been merged in and is now working.
